packages/gator-permissions-snap/src/profileSync/profileSync.ts

```diff
@@ -12,7 +13,82 @@ import type {
   JwtBearerAuth,
   UserStorage,
 } from '@metamask/profile-sync-controller/sdk';
-import { UnsupportedMethodError } from '@metamask/snaps-sdk';
+import {
+  InvalidInputError,
+  LimitExceededError,
+  ParseError,
+  UnsupportedMethodError,
+} from '@metamask/snaps-sdk';
+import { z } from 'zod';
+
+// Constants for validation
+const MAX_STORAGE_SIZE_BYTES = 400 * 1024; // 400kb limit as documented
+
+// Zod schema for runtime validation of StoredGrantedPermission
+const zStoredGrantedPermission = z.object({
+  permissionResponse: zPermissionResponse,
+  siteOrigin: z.string().min(1, 'Site origin cannot be empty'),
+});
+
+/**
+ * Safely deserializes and validates a JSON string as StoredGrantedPermission.
+ * @param jsonString - The JSON string to deserialize.
+ * @returns The validated StoredGrantedPermission object.
+ * @throws InvalidInputError if validation fails.
+ * @throws ParseError if parsing fails.
+ */
+function safeDeserializeStoredGrantedPermission(
+  jsonString: string,
+): StoredGrantedPermission {
+  try {
+    const parsed = JSON.parse(jsonString);
+    const validated = zStoredGrantedPermission.parse(parsed);
+    return validated;
+  } catch (error) {
+    logger.error('Error deserializing stored granted permission:', error);
+    if (error instanceof z.ZodError) {
+      throw new InvalidInputError(
+        `Invalid permission data structure: ${error.errors.map((zodError) => `${zodError.path.join('.')}: ${zodError.message}`).join(', ')}`,
+      );
+    }
+    throw new ParseError(`Failed to parse JSON`);
+  }
+}
+
+/**
+ * Safely serializes a StoredGrantedPermission object with size validation.
+ * @param permission - The permission object to serialize.
+ * @returns The JSON string.
+ * @throws LimitExceededError if size limit exceeded.
+ */
+function safeSerializeStoredGrantedPermission(
+  permission: StoredGrantedPermission,
+): string {
+  try {
+    // Validate the object structure first
+    const validated = zStoredGrantedPermission.parse(permission);
+
+    // Serialize to JSON
+    const jsonString = JSON.stringify(validated);
+
+    // Check size limit
+    const sizeBytes = new TextEncoder().encode(jsonString).length;
+    if (sizeBytes > MAX_STORAGE_SIZE_BYTES) {
+      throw new LimitExceededError(
+        `Permission data exceeds size limit: ${sizeBytes} bytes > ${MAX_STORAGE_SIZE_BYTES} bytes`,
+      );
+    }
+
+    return jsonString;
+  } catch (error) {
+    if (error instanceof z.ZodError) {
+      throw new InvalidInputError(
+        `Invalid permission data structure: ${error.errors.map((zodError) => `${zodError.path.join('.')}: ${zodError.message}`).join(', ')}`,
```

Comment:
I think we already have an `extractZodError` helper in `packages/shared/src/utils/error.ts` for handling Zod error messages.Perhaps we could reuse it here?